From 61b8d86697bc16b0c2d513d1e6e0bfc998231b4c Mon Sep 17 00:00:00 2001
From: Norbert Thiebaud <nthiebaud@gmail.com>
Date: Thu, 23 Jun 2011 20:08:16 -0500
Subject: [PATCH 4/9] move GIT_REPO_NAMES init earlier in the script

---
 configure.in |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/configure.in b/configure.in
index 77d2cb0..dd0c51b 100755
--- a/configure.in
+++ b/configure.in
@@ -1276,6 +1276,7 @@ AC_ARG_WITH(max-jobs,
 
 BUILD_TYPE="LibO"
 SCPDEFS=""
+GIT_REPO_NAMES="help"
 
 echo "********************************************************************"
 echo "*"
@@ -8517,7 +8518,6 @@ AC_SUBST(OOO_JUNIT_JAR)
 dnl ===================================================================
 dnl Dealing with l10n options
 dnl ===================================================================
-GIT_REPO_NAMES="help"
 AC_MSG_CHECKING([which languages to be built])
 # get list of all languages
 # generate shell variable from completelangiso= from solenv/inc/langlist.mk
-- 
1.7.3.4

