From d7a2f4e2210d665cd283dc836785187ecce5563a Mon Sep 17 00:00:00 2001
From: Norbert Thiebaud <nthiebaud@gmail.com>
Date: Thu, 21 Jul 2011 08:54:13 -0500
Subject: move GIT_REPO_NAMES init up

---
 configure.in |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/configure.in b/configure.in
index d400124..a6f2262 100755
--- a/configure.in
+++ b/configure.in
@@ -1276,6 +1276,7 @@ AC_ARG_WITH(max-jobs,
 
 BUILD_TYPE="LibO"
 SCPDEFS=""
+GIT_REPO_NAMES="help"
 
 echo "********************************************************************"
 echo "*"
@@ -8544,7 +8545,6 @@ AC_SUBST(OOO_JUNIT_JAR)
 dnl ===================================================================
 dnl Dealing with l10n options
 dnl ===================================================================
-GIT_REPO_NAMES="help"
 AC_MSG_CHECKING([which languages to be built])
 # get list of all languages
 # generate shell variable from completelangiso= from solenv/inc/langlist.mk
-- 
1.7.3.4

