From fdca59a4981a2ed599ccee240802d3c4266b0f18 Mon Sep 17 00:00:00 2001
From: Norbert Thiebaud <nthiebaud@gmail.com>
Date: Thu, 23 Jun 2011 20:09:43 -0500
Subject: [PATCH 5/9] binfilter is now in an optional repository

---
 configure.in |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/configure.in b/configure.in
index dd0c51b..2a06dd5 100755
--- a/configure.in
+++ b/configure.in
@@ -1979,6 +1979,7 @@ if test "$enable_binfilter" = "no"; then
 else
    WITH_BINFILTER="YES"
    BUILD_TYPE="$BUILD_TYPE BINFILTER"
+   GIT_REPO_NAMES="$GIT_REPO_NAMES binfilter"
    AC_MSG_RESULT([yes])
 fi
 AC_SUBST(WITH_BINFILTER)
-- 
1.7.3.4

