From b52306da67867036944bae2156e69a33ae6287a3 Mon Sep 17 00:00:00 2001
From: Norbert Thiebaud <nthiebaud@gmail.com>
Date: Thu, 23 Jun 2011 20:10:27 -0500
Subject: [PATCH 6/9] dictionaries is now in a separate optional repository

---
 configure.in |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/configure.in b/configure.in
index 2a06dd5..f0eaadb 100755
--- a/configure.in
+++ b/configure.in
@@ -2014,6 +2014,7 @@ if test -z "$with_myspell_dicts" || test "$with_myspell_dicts" = "yes"; then
   AC_MSG_RESULT([yes])
   WITH_MYSPELL_DICTS=YES
   BUILD_TYPE="$BUILD_TYPE DICTIONARIES"
+  GIT_REPO_NAMES="$GIT_REPO_NAMES dictionaries"
 else
   AC_MSG_RESULT([no])
   WITH_MYSPELL_DICTS=NO
-- 
1.7.3.4

