From 26e04d25df093c740a7ef6dbd9db5cef265c2d1e Mon Sep 17 00:00:00 2001
From: Norbert Thiebaud <nthiebaud@gmail.com>
Date: Thu, 21 Jul 2011 08:57:49 -0500
Subject: dictionaries is now in a separate optional repo

---
 configure.in |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/configure.in b/configure.in
index 8a1d319..b6ca892 100755
--- a/configure.in
+++ b/configure.in
@@ -2028,6 +2028,7 @@ if test -z "$with_myspell_dicts" || test "$with_myspell_dicts" = "yes"; then
     AC_MSG_RESULT([yes])
     WITH_MYSPELL_DICTS=YES
     BUILD_TYPE="$BUILD_TYPE DICTIONARIES"
+    GIT_REPO_NAMES="$GIT_REPO_NAMES dictionaries"
 else
     AC_MSG_RESULT([no])
     WITH_MYSPELL_DICTS=NO
-- 
1.7.3.4

