From 30bec927b0b755b26c9106462276038bdb69753a Mon Sep 17 00:00:00 2001
From: Jan Holesovsky <kendy@suse.cz>
Date: Mon, 18 Apr 2011 15:13:15 +0200
Subject: [PATCH 2/4] MSVC: Fix detection of the MSVC compiler.

---
 ccache.c |    6 +++++-
 1 files changed, 5 insertions(+), 1 deletions(-)

diff --git a/ccache.c b/ccache.c
index efede2a..b4e7e21 100644
--- a/ccache.c
+++ b/ccache.c
@@ -662,11 +662,15 @@ static void find_compiler(int argc, char **argv)
 	if (strcmp(base, MYNAME) == 0) {
 		args_remove_first(orig_args);
 		free(base);
+		base = str_basename(argv[1]);
+		if (strcmp(base, "cl") == 0 || strcmp(base, "cl.exe") == 0) {
+			compiler = COMPILER_MSVC;
+		}
 		if (strchr(argv[1],'/')) {
 			/* a full path was given */
+			free(base);
 			return;
 		}
-		base = str_basename(argv[1]);
 	}
 
 	/* support user override of the compiler */
-- 
1.7.2.3

