From cd5b0208db501364359fb53bd4f1d77dd58714bd Mon Sep 17 00:00:00 2001
From: Jan Holesovsky <kendy@suse.cz>
Date: Mon, 18 Apr 2011 16:12:24 +0200
Subject: [PATCH 3/4] MSVC: Re-route the preprocessor's stderr to stdout, to handle -showIncludes.

---
 ccache.c |   34 ++++++++++++++++++++++++++++++++--
 1 files changed, 32 insertions(+), 2 deletions(-)

diff --git a/ccache.c b/ccache.c
index b4e7e21..7df5d01 100644
--- a/ccache.c
+++ b/ccache.c
@@ -550,7 +550,8 @@ static void find_hash(ARGS *args)
 static void from_cache(int first)
 {
 	int fd_stderr, fd_cpp_stderr;
-	char *stderr_file;
+	int fd_stdout;
+	char *stderr_file, *stdout_file;
 	int ret;
 	struct stat st;
 
@@ -624,10 +625,39 @@ static void from_cache(int first)
 		i_tmpfile = NULL;
 	}
 
+	/* send the stdout - only on MSVC */
+	switch (compiler) {
+	case COMPILER_MSVC:
+		x_asprintf(&stdout_file, "%s.stdout", hashname);
+		fd_stdout = open(stdout_file, O_RDONLY | O_BINARY);
+		if (fd_stdout == -1) {
+			/* it isn't in cache ... */
+			free(stdout_file);
+			return;
+		}
+		copy_fd(fd_stdout, 1);
+		close(fd_stdout);
+		break;
+	default:
+		break;
+	}
+
 	/* send the cpp stderr, if applicable */
 	fd_cpp_stderr = open(cpp_stderr, O_RDONLY | O_BINARY);
 	if (fd_cpp_stderr != -1) {
-		copy_fd(fd_cpp_stderr, 2);
+		switch (compiler) {
+		case COMPILER_MSVC:
+			/* MSVC is confused, and writes the showIncludes to
+			   stderr when preprocessing, but to stdout when compiling.
+			   Here we assume that during preprocessing, only the
+			   dependency info is of some value, so let's store it
+			   where it is expected in the 'real run'. */
+			copy_fd(fd_cpp_stderr, 1);
+			break;
+		default:
+			copy_fd(fd_cpp_stderr, 2);
+			break;
+		}
 		close(fd_cpp_stderr);
 		unlink(cpp_stderr);
 		free(cpp_stderr);
-- 
1.7.2.3

